common/src/main/kotlin/streams/utils/ProcedureUtils.kt

```diff
@@ -0,0 +1,71 @@
+package streams.utils
+
+import org.neo4j.collection.RawIterator
+import org.neo4j.dbms.api.DatabaseManagementService
+import org.neo4j.exceptions.UnsatisfiedDependencyException
+import org.neo4j.kernel.internal.GraphDatabaseAPI
+import org.neo4j.logging.Log
+import org.neo4j.logging.internal.LogService
+import streams.extensions.execute
+import java.lang.invoke.MethodHandles
+import java.lang.invoke.MethodType
+
+
+object ProcedureUtils {
+    @JvmStatic
+    private val raftMachineClass: Class<*>? = try {
+        Class.forName("com.neo4j.causalclustering.core.consensus.RaftMachine")
+    } catch (e: ClassNotFoundException) {
+        null
+    }
+
+    @JvmStatic
+    private val isLeaderMethodHandle = raftMachineClass?.let {
+        val lookup = MethodHandles.lookup()
+        lookup.findVirtual(it, "isLeader", MethodType.methodType(Boolean::class.java))
+            .asType(MethodType.methodType(Boolean::class.java, Any::class.java))
+    }
+
+    fun clusterMemberRole(db: GraphDatabaseAPI): String {
+        val fallback: (Exception?) -> String = { e: Exception? ->
+            val userLog = db.dependencyResolver
+                .resolveDependency(LogService::class.java)
+                .getUserLog(ProcedureUtils::class.java)
+            e?.let { userLog.warn("Cannot call the APIs, trying with the cypher query", e) }
+                ?: userLog.warn("Cannot call the APIs, trying with the cypher query")
+            db.execute("CALL dbms.cluster.role(\$database)",
+                mapOf("database" to db.databaseName())
+            ) { it.columnAs<String>("role").next() }
+        }
+        val execute = {
+            raftMachineClass?.let {
+                try {
+                    val raftMachine: Any = db.dependencyResolver.resolveDependency(raftMachineClass)
```

Comment:
If we must use the dependency resolver to fetch something internal I'd rather we use a less powerful interface, like `LeaderLocator`, which will give you the `RaftMachine` but without most of the API. If you want to use `LeaderLocator#getLeaderInfo` you'll also need to resolve the `ClusteringIdentityModule` in order to compare `LeaderLocator#memberId()` to `ClusteringIdentityModule#memberId( id for this database )`.

Reply:
Sorry Hugo but it don't have access to Neo4j codebase so it would be good to have more info via email if it's possibile

Reply:
There is an interface which exists on all Neo4j versions from 3.5 on called `CoreMetaData`. 

The `RaftMachine` implements this interface. Its source is:

```
public interface CoreMetaData
{
    boolean isLeader();
}
```

If you use `DependencyResolver` to resolve `CoreMetaData` you will be getting the capability you need, without accessing `RaftMachine` itself 🙂